DTNL.LL.Logic/ProjectTimerService.cs

```diff
@@ -0,0 +1,102 @@
+        {
+            _tickTime = options.Value.TickTimeInSeconds;
+        }
+
+        /// <summary>
+        /// This method updates the internal list of projects that go into sleep mode. This should be called before GetTickedProjects as that method will filter out all sleeping methods.
+        /// </summary>
+        /// <param name="projects">A list of all projects that should be updated in the internal database.</param>
+        /// <returns>A list of all projects that will go to sleep mode this tick. Lamps in this list should be turned off.</returns>
+        public List<Project> UpdateSleepingProjectList(IEnumerable<Project> projects)
+        {
+            var newSleepingProjects = new List<Project>();
+            foreach (var project in projects)
+            {
+                if (!project.TimeRangeEnabled)
+                {
+                    // In case a setting has changed, make sure project is not in the disabled list.
+                    if (_sleepingProjectIds.Contains(project.Id))
+                        _sleepingProjectIds.Remove(project.Id);
+                    continue;
+                }
+
+                var awake = IsTimeOfDayBetween(DateTime.Now, project.TimeRangeStart!.Value, project.TimeRangeEnd!.Value);
+
+                if (!awake)
+                {
+                    if (!_sleepingProjectIds.Contains(project.Id))
+                        newSleepingProjects.Add(project);
+
+                    _sleepingProjectIds.Add(project.Id);
+                    continue;
+                }
+
+                if(_sleepingProjectIds.Contains(project.Id))
+                    _sleepingProjectIds.Remove(project.Id);
+            }
+
+            return newSleepingProjects;
+        }
+
+        /// <summary>
+        /// Returns a list with the projects that should be polled this tick.
+        /// Each time this method is called the internal timer will be raised with set amount of seconds in the ServiceWorker options.
+        /// </summary>
+        /// <param name="projects"></param>
+        /// <returns>Returns a list of project whose light and analytics should updated.</returns>
+        public List<Project> GetTickedProjects(IEnumerable<Project> projects)
+        {
+            _secondsPassed += _tickTime;
+
+            // Make sure an overflow does not happen.
+            _secondsPassed = Math.Max(0, _secondsPassed);
+            var tickedProjects = new List<Project>();
+
+            foreach (var project in projects)
+            {
+                var pollingTimeInSeconds = project.PollingTimeInMinutes * SecondsPerMinute;
+                if(_secondsPassed % pollingTimeInSeconds != 0)
+                    continue;
+
+                tickedProjects.Add(project);
+            }
+
+            return tickedProjects;
+        }
+
+        public static bool IsTimeOfDayBetween(DateTime time,
+            TimeSpan startTime, TimeSpan endTime)
+        {
+            if (endTime == startTime)
```

Comment:
This is strange,
so if someone configure the same start and end time, it is always active?

Reply:
The original implementation I wrote didn't use a seperate boolean that would signify always active but just set the times to the same values. So it would stay always active if the times were the same. I left it like this just in case I would need the method again.

Reply:
Alright, so this is a test case. You should remove it for production.

Because we should also turn the lamp on at starttime and off at endtime. So if both are on the same time. Then this is going to be a problem.